lib/compress/zstd_fast.c

```diff
@@ -183,6 +183,237 @@ ZSTD_compressBlock_fast_generic(
+    const U32   endIndex = (U32)((size_t)(istart - base) + srcSize);
+    const U32   prefixStartIndex = ZSTD_getLowestPrefixIndex(ms, endIndex, cParams->windowLog);
+    const BYTE* const prefixStart = base + prefixStartIndex;
+    const BYTE* const iend = istart + srcSize;
+    const BYTE* const ilimit = iend - HASH_READ_SIZE;
+
+    const BYTE* anchor = istart;
+    const BYTE* ip0 = istart;
+    const BYTE* ip1;
+    const BYTE* ip2;
+    U32 current0;
+
+    U32 rep_offset1 = rep[0];
+    U32 rep_offset2 = rep[1];
+    U32 offsetSaved = 0;
+
+    size_t hash0; /* hash for ip0 */
+    size_t hash1; /* hash for ip1 */
+    U32 idx; /* match idx for ip0 */
+    U32 mval; /* src value at match idx */
+
+    U32 offcode;
+    const BYTE* match0;
+    size_t mLength;
+
+    size_t step;
+    const BYTE* nextStep;
+    const size_t kStepIncr = (1 << (kSearchStrength - 1));
+
+    DEBUGLOG(5, "ZSTD_compressBlock_fast_generic_pipelined");
+    ip0 += (ip0 == prefixStart);
+    {   U32 const curr = (U32)(ip0 - base);
+        U32 const windowLow = ZSTD_getLowestPrefixIndex(ms, curr, cParams->windowLog);
+        U32 const maxRep = curr - windowLow;
+        if (rep_offset2 > maxRep) offsetSaved = rep_offset2, rep_offset2 = 0;
+        if (rep_offset1 > maxRep) offsetSaved = rep_offset1, rep_offset1 = 0;
+    }
+
+    /* start each op */
+_start: /* Requires: ip0 */
+
+    step = stepSize;
+    nextStep = ip0 + kStepIncr;
+
+    /* calculate positions, ip0 - anchor == 0, so we skip step calc */
+    ip1 = ip0 + stepSize;
+    ip2 = ip1 + stepSize;
+
+    if (ip2 >= ilimit) {
+        goto _cleanup;
+    }
+
+    hash0 = ZSTD_hashPtr(ip0, hlog, mls);
+    hash1 = ZSTD_hashPtr(ip1, hlog, mls);
+
+    idx = hashTable[hash0];
+
+    do {
+        const U32 rval = MEM_read32(ip2 - rep_offset1);
+        current0 = ip0 - base;
+
+        /* write back hash table entry */
+        hashTable[hash0] = current0;
+
+        /* check repcode at ip[2] */
+        if ((MEM_read32(ip2) == rval) & (rep_offset1 > 0)) {
+            ip0 = ip2;
+            match0 = ip0 - rep_offset1;
+            mLength = ip0[-1] == match0[-1];
+            ip0 -= mLength;
+            match0 -= mLength;
+            offcode = 0;
+            mLength += 4;
+            goto _match;
+        }
+
+        /* load match for ip[0] */
+        if (idx >= prefixStartIndex) {
+            mval = MEM_read32(base + idx);
+        } else {
+            mval = MEM_read32(ip0) ^ 1; /* guaranteed to not match. */
+        }
+
+        /* check match at ip[0] */
+        if (MEM_read32(ip0) == mval) {
+            /* found a match! */
+            goto _offset;
+        }
+
+        hash0 = hash1;
+
+        /* hash ip[2] */
+        hash1 = ZSTD_hashPtr(ip2, hlog, mls);
+
+        /* lookup ip[1] */
+        idx = hashTable[hash0];
+
+        /* advance to next positions */
+        {
+            if (ip1 >= nextStep) {
+                PREFETCH_L1(ip1 + 64);
```

Comment:
why `ip1 + 64` ?

Reply:
Also : why no `prefetch` when `ip1 < nextStep` ?
This would essentially never prefetch when the distance between consecutive matches remain small ?

Reply:
This fetches the next cache line of the input.

This is a very rough heuristic. Prefetching on every loop iter is too much and slows the loop down. So the advantage of tucking a prefetch into this branch is that: (a) it's not in the hot loop, (b) it happens less frequently, and (c) it happens when we're operating over an incompressible section where increasing our lookahead into the input is most beneficial.

It actually looks like also fetching `ip1 + 128` is a very slight win, so I'll add that. Since we take this branch every 128 bytes, I guess that makes sense--we'll prefetch every upcoming cache line of the input exactly once when we hit a big incompressible section.